### PR TITLE
Use boost::regex as a fallback on CentOS 7

### LINKED ISSUE #800: CentOS build failing after merging #796
After we merged #796 , some tests started to fail in the CentOS build. @D4N sensed that this could happen and he might have some intuition of why this situation has happened. 

Update after investigation:
The problem comes from the fact that CentOS has a very old very of GCC in which regex is not  available:
```
--> Processing Dependency: gcc = 4.8.5-36.el7_6.1 for package: gcc-c++-4.8.5-36.el7_6.1.x86_64
```

See this [link](https://stackoverflow.com/a/12665408/1977218) for more info.

Possible approaches to deal with this situation:
- Upgrade CentOS on the CI ? I am not sure if were are using the latest version or not.
- Remove CentOS from the Gitlab pipelines, since the GCC version it contains it is not fully C++11 compatible.

### PR DESCRIPTION
**Pull request by D4N**
I've added a check to our cmake code that verifies whether the implementation of `<regex>` is supposed to work. If it does not (currently only known for CentOS/RHEL 7), then we fall back to use boost::regex instead.

This fixes #800.

### CODE REVIEW

**Comment by D4N** [comment]
Needs fixing on Windows.

File: cmake/findDependencies.cmake [hunk 1]
```diff
@@ -59,3 +59,30 @@ if( BUILD_WITH_CCACHE )
...
+             _GLIBCXX_RELEASE > 4)))
+int main() { return 0; }
+#else
+#error \"regex not working\"
+#endif
+"
+  BUILTIN_REGEX_WORKING
+  FAIL_REGEX '.*regex not working.*'
+  )
+set(CMAKE_REQUIRED_FLAGS "")
```
Comment (piponazo) [inline cmake/findDependencies.cmake]: [minor] According to the CMake doc, `CMAKE_REQUIRED_FLAGS` only affects to the `check_cxx_source_compiles` and `check_c_source_compiles` commands. Therefore, I think we do not need to clear the variable here. 

https://cmake.org/cmake/help/latest/module/CheckCXXSourceCompiles.html?highlight=checkcxxsourcecompiles

File: cmake/findDependencies.cmake [hunk 2]
```diff
@@ -59,3 +59,30 @@ if( BUILD_WITH_CCACHE )
...
+
+if( NOT BUILTIN_REGEX_WORKING )
+  set( EXV_NEED_BOOST_REGEX ON
+    CACHE
+    BOOL
+    "Need Boost::regex as this version of gcc ships a broken implementation of <regex>")
+  find_package(Boost REQUIRED COMPONENTS regex)
```
Comment (piponazo) [inline cmake/findDependencies.cmake]: Probably we should specify here the minimum version required.

File: README.md [hunk 3]
```diff
@@ -768,6 +769,17 @@ $ cmake .. -G "Unix Makefiles"
 $ make
 ```
 
+#### 5.1.1 CentOS/RHEL 7
+
+CentOS/RHEL 7 comes with gcc 4.8, which is compatible with C++11, but unfortanely ships a broken implementation of `<regex>`.
+We work around this issue by falling back to using `boost::regex` in this case.
+
+To build exiv2 on CentOS, please install the following additional packages:
+- `cmake3`
```
Comment (piponazo) [inline README.md]: I do not think the cmake package should be listed here. It is already a dependency for building the project, even if boost is not going to be used.

**Review by piponazo — DISMISSED** [review]
Overall, LGTM. Thanks!

I just made few minor suggestions.

File: cmake/findDependencies.cmake (hunk 1 above)
Comment (D4N) [reply cmake/findDependencies.cmake]: Unfortunately it is required to clear it. I initially did not do that and the XMPSDK would suddenly no longer compiler, clearing it fixed the issue. Don't ask me why…

File: README.md (hunk 3 above)
Comment (D4N) [reply README.md]: I added it for completeness, as the `cmake` package is cmake 2.8, but we require cmake 3.

File: cmake/findDependencies.cmake (hunk 2 above)
Comment (D4N) [reply cmake/findDependencies.cmake]: Hm, to be honest I wouldn't know which version to choose for the minimum. I could insert the currently available one from EPEL, but no idea if older ones would work too?

**Comment by D4N** [comment]
If the CI passes then I need to rebase the fixup comment away, so please no merging before that ;-)

**Comment by D4N** [comment]
> If the CI passes then I need to rebase the fixup comment away, so please no merging before that ;-)

Rebased.

**Review by piponazo — APPROVED** [review]

**Pull request merged**